File: js/day11.js
const Dir = {
  Top: 0,
  Right: 1,
  Bottom: 2,
  Left: 3
};

const Rotation = {
  Anticlockwise: 0,
  Clockwise: 1
};

const Color = {
  Black: 0,
  White: 1
};

function rotate(rotation, dir) {
  if (rotation === Rotation.Clockwise) {
    switch (dir) {
      case Dir.Top:
        return Dir.Right;
      case Dir.Right:
        return Dir.Bottom;
      case Dir.Bottom:
        return Dir.Left;
      case Dir.Left:
        return Dir.Top;
    }
  }

  switch (dir) {
    case Dir.Top:
      return Dir.Left;
    case Dir.Left:
      return Dir.Bottom;
    case Dir.Bottom:
      return Dir.Right;
    case Dir.Right:
      return Dir.Top;
  }
}

class Grid {
  constructor() {
    this._hash = {};
    this._minX = 0;
    this._maxX = 0;
    this._minY = 0;
    this._maxY = 0;
  }

  get coloredCount() {
    let result = 0;
    for (let y in this._hash) {
      for (let _ in this._hash[y]) {
        result++;
      }
    }
    return result;
  }

  get corners() {
    return {
      topLeft: { x: this._minX, y: this._minY },
      bottomRight: { x: this._maxX, y: this._maxY }
    };
  }

  set(pos, color) {
    this.updateMinMax(pos);
    this._hash[pos.y] = this._hash[pos.y] || {};
    this._hash[pos.y][pos.x] = color;
    return this;
  }

  get(pos) {
    return (this._hash[pos.y] && this._hash[pos.y][pos.x]) || Color.Black;
  }

  updateMinMax(pos) {
    this._minX = Math.min(this._minX, pos.x);
    this._maxX = Math.max(this._maxX, pos.x);
    this._minY = Math.min(this._minY, pos.y);
    this._maxY = Math.max(this._maxY, pos.y);
  }
}

class Robot {
  constructor(registries) {
    this._program = require("./IntcodeComputer").program(registries);
    this._pos = { x: 0, y: 0 };
    this._dir = Dir.Top;
  }

  get position() {
    return this._pos;
  }

  get direction() {
    return this._dir;
  }

  isHalted() {
    return this._program.isHalted();
  }

  /** @param {Grid} grid */
  paint(grid, color) {
    grid.set(this._pos, color);
    return this;
  }

  /** @param {Grid} grid */
  step(grid) {
    const input = [grid.get(this._pos)];
    const { output } = this._program.pushInput(input).run();
    const [color, rotation] = output.slice(-2);
    return this.paint(grid, color)
      .rotate(rotation)
      .stepForward();
  }

  rotate(rotation) {
    this._dir = rotate(rotation, this._dir);
    return this;
  }

  stepForward() {
    switch (this._dir) {
      case Dir.Top:
        this._pos.y -= 1;
        break;
      case Dir.Right:
        this._pos.x += 1;
        break;
      case Dir.Bottom:
        this._pos.y += 1;
        break;
      case Dir.Left:
        this._pos.x -= 1;
        break;
    }
    return this;
  }
}

function solve(registries, initColor) {
  const robot = new Robot(registries);
  const grid = new Grid().set(robot.position, initColor);
  while (!robot.isHalted()) {
    robot.step(grid);
  }
  return grid;
}

/**
 * @param {Grid} grid
 * @param {Robot} robot
 */
function drawGrid(grid, robot) {
  const { bottomRight, topLeft } = grid.corners;
  let result = "";
  for (let y = topLeft.y - 1; y <= bottomRight.y + 1; y++) {
    let line = "";
    for (let x = topLeft.x - 1; x <= bottomRight.x + 1; x++) {
      if (robot.position.x === x && robot.position.y === y) {
        line += drawDirection(robot.direction) + " ";
      } else {
        line += drawColor(grid.get({ x, y })) + " ";
      }
    }
    result += line + "\n";
  }
  return result;
}

function drawDirection(direction) {
  switch (direction) {
    case Dir.Top:
      return "👆";
    case Dir.Right:
      return "👉";
    case Dir.Bottom:
      return "👇";
    case Dir.Left:
      return "👈";
  }
}

function drawColor(color) {
  if (color === Color.White) {
    return "🌺";
  }
  return "🌱";
}

function clear() {
  process.stdout.cursorTo(0, 0);
  process.stdout.clearScreenDown();
}

function solveAnimated(registries, initColor, speed) {
  const robot = new Robot(registries);
  const grid = new Grid().set(robot.position, initColor);

  drawGrid(grid, robot);

  function step(resolve) {
    robot.step(grid);
    clear();
    process.stdout.write(drawGrid(grid, robot));

    if (robot.isHalted()) {
      resolve();
    } else {
      setTimeout(() => {
        step(resolve);
      }, speed);
    }
  }

  return new Promise(step);
}

require("./Problem")({
  input() {
    return require("./Input")
      .day(11)
      .split(",")
      .map(Number);
  },
  solve(registries) {
    return solveAnimated(registries, Color.White, 16);
  }
}).run();

File: haskell/src/Day11.hs
module Day11
    ( part1
    , part2
    )
where

import           Lib
import           Data.List
import qualified Data.Map                      as M
import           IntcodeComputer
import           Data.Maybe

prepareInput :: IO Program
prepareInput =
    fromRegistries . map (read :: String -> Int) <$> readInputForDayAsCSV 11

part1 = do
    p <- prepareInput
    let (dir, grid, pos, colored) = buildHull p { input = [0] }
    putStr (drawMap pos dir grid)
    return $ M.size colored

part2 = do
    p <- prepareInput
    let (dir, grid, pos, _) = buildHull p { input = [1] }
    putStr (drawMap pos dir grid)
    return ()

type Grid = M.Map Pos Color

type Colored = M.Map Pos Int

type Pos = (Int, Int)

type Step = (Color, Rotation)

data Dir = TopD | RightD | BottomD | LeftD deriving (Show, Eq)

data Rotation = AntiClockwise | Clockwise deriving (Enum, Show, Eq)

data Color = Black | White deriving (Enum, Show, Eq)

type HullResult = (Dir, Grid, Pos, Colored)

rotate :: Rotation -> Dir -> Dir
rotate Clockwise dir = case dir of
    TopD    -> RightD
    RightD  -> BottomD
    BottomD -> LeftD
    LeftD   -> TopD
rotate AntiClockwise dir = case dir of
    TopD    -> LeftD
    LeftD   -> BottomD
    BottomD -> RightD
    RightD  -> TopD

toRotation :: Int -> Rotation
toRotation = toEnum

toColor :: Int -> Color
toColor = toEnum

getNextStep :: Program -> Step
getNextStep =
    (\(d : c : _) -> (toColor c, toRotation d)) . take 2 . reverse . output

moveTo :: Dir -> Pos -> Pos
moveTo TopD    (x, y) = (x, y - 1)
moveTo BottomD (x, y) = (x, y + 1)
moveTo LeftD   (x, y) = (x - 1, y)
moveTo RightD  (x, y) = (x + 1, y)

applyStep :: Step -> HullResult -> HullResult
applyStep (color, rotation) (dir, grid, pos, coloredCells) =
    (dir', grid', pos', coloredCells')
  where
    grid'         = M.insert pos color grid
    coloredCells' = M.insertWith (+) pos 1 coloredCells
    dir'          = rotate rotation dir
    pos'          = moveTo dir' pos

buildHull :: Program -> HullResult
buildHull = curry buildHull' (TopD, M.fromList [], (0, 0), M.fromList [])
  where
    buildHull' (result, p) | isHalted p = result
                           | otherwise  = buildHull' $ buildNextHull (result, p)

buildNextHull :: (HullResult, Program) -> (HullResult, Program)
buildNextHull (result, p) = (result', p'')
  where
    p'                        = runProgramm p
    step                      = getNextStep p'
    result'@(_, grid, pos, _) = applyStep step result
    curColor                  = fromMaybe Black (M.lookup pos grid)
    p''                       = p' { input = input p' ++ [fromEnum curColor] }

drawMap :: Pos -> Dir -> Grid -> String
drawMap pos dir grid = unlines image
  where
    allCoords = M.keys grid
    xs        = map fst allCoords
    ys        = map snd allCoords
    minX      = min (minimum xs - 2) (-5)
    maxX      = max (maximum xs + 2) 5
    minY      = min (minimum ys - 2) (-5)
    maxY      = max (maximum ys + 2) 5
    width     = maxX - minX
    height    = maxY - minY
    getChar' pos' = if pos == pos' then drawDir dir else getMapChar pos' grid
    drawLine y = [ getChar' (x, y) | x <- [minX .. maxX] ]
    image = [ drawLine y | y <- [minY .. maxY] ]

getMapChar :: Pos -> Grid -> Char
getMapChar pos grid = maybe ' ' colorToCh (M.lookup pos grid)
  where
    colorToCh White = 'O'
    colorToCh Black = ' '

drawDir :: Dir -> Char
drawDir dir = case dir of
    TopD    -> '˄'
    RightD  -> '˃'
    BottomD -> '˅'
    LeftD   -> '˂'

